### PR TITLE
feat : Add payments to the App

### PR DESCRIPTION
**Pull request by simarsingh24**
first steps to #237

### CODE REVIEW

**Comment by nikit19** [comment]
Can you post your request that you see in the logs

File: app/src/main/res/layout/fragment_attendee.xml [hunk 1]
```diff
@@ -74,6 +83,39 @@
...
+        android:layout_marginTop="@dimen/padding_medium">
+
+        <EditText
+            android:id="@+id/country"
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content"
+            android:hint="Country"/>
+    </android.support.design.widget.TextInputLayout>
+
+
```
Comment (dreadpool2) [inline app/src/main/res/layout/fragment_attendee.xml]: extra line

**Comment by simarsingh24** [comment]
@iamareebjamal @nikit19 @dreadpool2 I am sharing the logs please take a look at it ,
```
POST https://open-event-api-dev.herokuapp.com/v1/orders?include=event,attendees&fields[event]=id&fields[attendees]=id
    Content-Type: application/vnd.api+json
    Content-Length: 258
  {
  	"data": {
  		"type": "order",
  		"id": "40",
  		"attributes": {
  			"payment-mode": "paypal",
  			"country": "india",
  			"status": "pending",
  			"amount": 2.0,
  			"order-notes": "adf"
  		},
  		"relationships": {
  			"attendee": {
  				"data": [{
  					"type": "attendee",
  					"id": "40"
  				}]
  			},
  			"event": {
  				"data": {
  					"type": "event",
  					"id": "68"
  				}
  			}
  		}
  	}
  }
    --> END POST (258-byte body)
```
```
{
	"errors": [{
		"status": 422,
		"source": {
			"pointer": "/data/relationships/ticket_holders"
		},
		"title": "Unprocessable Entity",
		"detail": "A valid relationship with ticket_holders resource is required"
	}],
	"jsonapi": {
		"version": "1.0"
	}
}
```

**Comment by iamareebjamal** [comment]
Why are you sending ID?

**Comment by srv-twry** [comment]
@simarsingh24 Two points:
1. How do you have the `id` when you're actually sending a POST request ?
2. Orders endpoint is a really complicated endpoint to work with and you've to handle a lot of cases.

Here are the steps that are followed by the FE:
1. User selects the tickets for the order.
2. We create order/attendees on the FE. Push the attendees(with random names or blank names) so that tickets are locked. Push the order on the server with the relationship with attendees and event to get the hash/identifier.
3. Update the names to blank on the FE(don't push to server) <- **No need for this step if you had blank names for attendees.**
4. User get's the form, and 10 minutes slot. If he/she fills it, great. We update the order with the status "placed" and other details and start the payment procedure using the charges api. Also update the attendee names. After the payment succeeds we update the status to "completed" and send tickets/notifications otherwise we set the status to "expired". We send back the error. Show the payment error page with the info about the error(v1 does it).
5. If they weren't able to enter the info, simply expire the order and show them an alert that the order has expired and redirect to the public page(v1 does it) or anywhere else.

Similar to this procedure, here is a rough estimate of what you can do in the app:

1. User selects the tickets for the order.
2. We create order/attendees on the app. Push the attendees(with random names or blank names) so that tickets are locked. Push the order on the server with the relationship with attendees and event to get the hash/identifier.
3. Update the names to blank on the app(don't push to server) <- **No need for this step if you had blank names for attendees.**
4. User get's the form, and 10 minutes slot. If he/she fills it, great. Make a PATCH request to the orders endpoint with the updated details and then call the charges API. It is probably not added in the docs but will be added soon. Also update the attendee names by making a PATCH request. After the payment succeeds we update the order status to "completed" and send tickets/notifications otherwise we set the status to "expired". We send back the error to the app. Show the error to the user.
5. If they weren't able to enter the info in 10 minutes, simply expire the order and show them an alert that the order has expired and redirect to anywhere you see fit.

Of course you can see how it is done in the Eventbrite general app and make minor adjustments.

**I would recommend first discussing the whole workflow, adding it in the issue and then implementing it.**

**Comment by simarsingh24** [comment]
@iamareebjamal  checked sending same request without id on postman still the error response is same.
@srv-twry Can you please confirm the documentation for this is according to the latest update

**Comment by srv-twry** [comment]
It should be `attendees` not `attendee`.
```
"attendees": {
        "data": [
          {
            "id": "1",
            "type": "attendee"
          }
        ]
      },
```

**Comment by srv-twry** [comment]
Sending this POST request works for me:
```
{
  "data": {
    "attributes": {
      "payment-mode": "free",
      "country": "India",
      "status": "pending",
      "amount": "100"
    },
    "type": "order",
    "relationships": {
      "attendees": {
        "data": [
          {
            "id": "1",
            "type": "attendee"
          }
        ]
      },
      "event": {
        "data": {
          "id": "{{event}}",
          "type": "event"
        }
      }
    }
  }
}
```

Match yours with this.

**Comment by simarsingh24** [comment]
@srv-twry @iamareebjamal I created an attendee logged his id and made a request using postman i got
```
{
    "errors": [
        {
            "status": 409,
            "source": {
                "pointer": "/data/relationships/attendees"
            },
            "title": "Conflict",
            "detail": "Attendee with id 316 does not exists"
        }
    ],
    "jsonapi": {
        "version": "1.0"
    }
}
```

**Comment by simarsingh24** [comment]
created attendee through this app only

**Comment by iamareebjamal** [comment]
`?include=event,attendees&fields[event]=id&fields[attendees]=id` This does not work with POST

**Comment by simarsingh24** [comment]
@iamareebjamal changed it, but the error still remains the same

**Comment by srv-twry** [comment]
> "id": "316",

How do you have the `id` of the order that isn't even created ?

**Comment by nikit19** [comment]
@srv-twry This value of the `id` is the user id extracted from the JWT token so that it is unique. This is required by ROOM as a primary key. The response will be the same if we add this `id` to the request or not so that should not be a problem

**Comment by srv-twry** [comment]
@nikit19 
> The response will be the same if we add this id to the request or not so that should not be a problem

Pardon, but I didn't understand what you meant. For every model in the server, id is the field which is the primary key of that model. If you're passing in the `id` for an order, then the server associates it with the order having id = 316 not the user with id = 316.

Regarding the POST request, the server doesn't even know what to do with the `id`. You post the data, ID Is generated inside the server and sent back to you as a response.
you have to send everything in accordance with the docs, nothing extra. It will never work.

**Comment by nikit19** [comment]
@srv-twry when I sent this PR https://github.com/fossasia/open-event-android/pull/206 to create an attendee. It successfully creates the attendee with and without the id.

**Comment by srv-twry** [comment]
> It successfully creates the attendee with and without the id.

May be then it silently overrides it inside the server without raising the error.
See, let's say 50 orders were created in the database on the server. Now when you post a new order, id=51 is assigned to this new order. Remove it from the response and try again please @simarsingh24

**Comment by srv-twry** [comment]
screenshot from 2018-07-04 15-49-26


Server is working properly.

**Comment by simarsingh24** [comment]
@srv-twry Are you doing it on you server setup locally or the deployed one? I tried making request using postman using  still getting server error and also in the shared screen shot you are not sending order-notes which if not sent gives `bad request` error.
Please check and revert back

**Comment by srv-twry** [comment]
@simarsingh24 I used the open-event-api-dev.herokuapp.com server (Check top right of the screenshot).
My post request works with and without sending the order notes. 

screenshot from 2018-07-05 07-07-36

**Comment by srv-twry** [comment]
@simarsingh24 Please post the screenshot of your postman request with the error response.

**Comment by simarsingh24** [comment]
```
 {
 	"data": {
 		"type": "order",
 		"attributes": {
 			"payment-mode": "paypal",
 			"country": "adfad",
 			"status": "pending",
 			"amount": 2.0,
 			"order-notes": "adfa"
 		},
 		"relationships": {
 			"attendees": {
 				"data": [{
 					"type": "attendee",
 					"id": "122"
 				}]
 			},
 			"event": {
 				"data": {
 					"type": "event",
 					"id": "68"
 				}
 			}
 		}
 	}
 }
```
error log
```
{
    "errors": [
        {
            "detail": "Unknown error",
            "source": {
                "pointer": ""
            },
            "status": 500,
            "title": "Unknown error"
        }
    ],
    "jsonapi": {
        "version": "1.0"
    }
}
```

**Comment by simarsingh24** [comment]
@srv-twry @iamareebjamal  found the error , if it make the payment mode from paypal to free it works, maybe its because the email id is not regestered on paypal.

File: app/src/main/res/layout/fragment_attendee.xml [hunk 2]
```diff
@@ -85,6 +127,10 @@
         android:padding="@dimen/padding_medium"
         android:text="@string/register"
         android:textColor="@android:color/white" />
+        </LinearLayout>
+
+    </android.support.v4.widget.NestedScrollView>
+
```
Comment (nikit19) [inline app/src/main/res/layout/fragment_attendee.xml]: Extra

File: app/src/main/res/layout/fragment_attendee.xml (hunk 1 above)
Comment (nikit19) [inline app/src/main/res/layout/fragment_attendee.xml]: extra

File: app/src/main/java/org/fossasia/openevent/general/attendees/AttendeeFragment.kt [hunk 3]
```diff
@@ -51,6 +53,19 @@ class AttendeeFragment : Fragment() {
...
+        paymentOptions.add("stripe")
+        rootView.paymentSelector.adapter = ArrayAdapter(context, android.R.layout.simple_spinner_dropdown_item, paymentOptions)
+        rootView.paymentSelector.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
+            override fun onNothingSelected(p0: AdapterView<*>?) {
+            }
+
+            override fun onItemSelected(p0: AdapterView<*>?, p1: View?, p2: Int, p3: Long) {
+                selectedPaymentOption = paymentOptions[p2]
+            }
+
```
Comment (nikit19) [inline app/src/main/java/org/fossasia/openevent/general/attendees/AttendeeFragment.kt]: extra line

**Comment by simarsingh24** [comment]
@nikit19 SS-
image

**Comment by simarsingh24** [comment]
@srv-twry done

**Comment by simarsingh24** [comment]
@iamareebjamal @nikit19 @dreadpool2 this is a blocker for further payments related works, please review and merge this

File: app/src/main/java/org/fossasia/openevent/general/attendees/AttendeeViewModel.kt [hunk 4]
```diff
@@ -40,14 +42,53 @@ class AttendeeViewModel(private val attendeeService: AttendeeService, private va
                 }.doFinally {
                     progress.value = false
                 }.subscribe({
+                    if (attendee.ticket?.id != null) {
+                        loadTicket(attendee.ticket!!.id, country, eventId, paymentOption, it.id)
```
[the remark this answers is not in the document]
Comment (dreadpool2) [reply app/src/main/java/org/fossasia/openevent/general/attendees/AttendeeViewModel.kt]: Please avoid non null assertion

**Comment by simarsingh24** [comment]
@iamareebjamal please review

**Pull request merged**